### ISSUE TITLE
/api/list/* - List API requirement (GET)

### ISSUE THREAD

**Issue by stemmlerjs** [UnivJobs/univjobs-back#81]
Hey guys, 
We discussed earlier the need for an **/api/list/*** routes as a way to retrieve all of the items within dropdowns. First, let's talk about the GET requirements.

**GET APIs**
- /api/list/emailpreferences 
- /api/list/studentstatuses
- /api/list/edulevel
- /api/list/majors
- /api/list/genders
- /api/list/sportsteams
- /api/list/schoolclubs
- /api/list/languages
- /api/list/cities
- /api/list/industries

**Return object format:**
Let's return the objects with their unique id and their text. It's important that we include this unique way for us to identify the selected item so that when we perform our PUT to /api/me (or some other endpoint in the future), we can easily create the relationship between our USERS table and our list (ie: CITIES_LIST, MAJORS_LIST, SPORTS_TEAMS_LIST) table.

Example: **/api/list/cities**
```javascript
{
   {
     id: 1,
     city: 'Mississauaga'
   },
   {
     id: 2,
     city: 'Oakville'
   }
}
```

Example: **/api/list/sportsteams**
```javascript
{
   {
     id: 1,
     team: 'Brock Badgers'
   },
   {
     id: 2,
     team: 'Connecticut Huskies'
   }
}
```

In summary, what needs to be done here is:

1. Create the corresponding table for each of these lists.
2. Create the corresponding GET api to retrieve each of the items in these lists formatted in (id, text) format.

**Comment by charlesakio** [comment]
@stemmlerjs , @juliosueiras  is making the api list different than what you asked for.
It will only have one endpoint which would look like this:

Example: **api/list/?q=major**
```
{0: Computer Science}
{1: Computer Engineer}
'''
'''
{99: Marine Biologist}
```
Example: **api/list/?q=industry**
```
{0: Aviation}
{1: Military}
'''
'''
{99: Sports and Entertainment}
```

You will type in the type of dropdown choice you want and it will return you all the you need.

**Comment by charlesakio** [comment]
@juliosueiras 

# TO DO:
- [x]  Finish the api/list request
- [ ]  Document the api/list request into apiary

**Comment by stemmlerjs** [comment]
I think it was discussed on the Google Docs and I should have included this in the requirement that I wrote here but did you guys write the /list APIs with an optional *query* parameter?

IE: I'm typing into the Industries field the Industry that my company belongs in. Looking for the industry titled "Telecommunicatons", as I type, the client side application is making requests to ```**GET /api/list/industries**?query=telecommun``` where the query parameter is what I've typed in.

This should essentially return to an array of objects that ran through a sql LIKE (such as **SELECT * FROM INDUSTRIES WHERE NAME LIKE ?**).

If this was assumed while you guys were making the endpoints, fantastic.
If not, it's what you should be doing next.

**Comment by stemmlerjs** [comment]
Thinking about this.
It's probably a more efficient to just run array.map() over all of the collected data than to make contiguous requests onKeypress.

**OK, what I asked for above is now a _nice-to-have_. I've decided I'll rerender the list onKeypress myself.**

**Comment by charlesakio** [comment]
@stemmlerjs, we can have this in the future. 

**First, here are the reasons as to why I would say no for now, though I am open to it in the future.**
- It will deter us from finishing other features.
- According to Julio this will stress at the backend. With each keypress with n numbers of users, the backend is not fit YET to handle the load.
- Another thing is that we do not know what people prefer **DROPDOWN OR FUZZY SEARCH?**


**How we make the call at the back**
One reason why is that this query is not going through the database. Rather just into a file called choices.py. The biggest reason is the time constraint, 

Here is an [example of students choices](https://github.com/UnivJobs/univjobs-back/blob/develop/univjobs/student/choices.py)

It goes through the [utils/views.py](https://github.com/UnivJobs/univjobs-back/blob/develop/univjobs/utils/views.py) to turn all the choices.py from each django app into JSON format.

With that said, I prefer to have onKeypress, but with time constraint. We can leave this at the end. When we finish every feature we set out to have.


**THINKING ABOUT USERS USING DATA**
    This is the part where we as a team have to think what the user feels more intuitive. Another     saying is we do not want them to think when they use our product. It has to feel natural. How we do that is through **data driven decisions** by doing some A/B testing.

**SUGGESTIONS**
     We can do a few A/B testings when we finish all the said features.
    
- Create two front-ends with the said features(fuzzy search and dropdown) with the least amount of implementation, given that you still have to parse the api list at the front.

**FUZZY SEARCH STEPS**
- Fuzzy search implementation steps using [HTML5 DataList](http://blog.teamtreehouse.com/creating-autocomplete-dropdowns-datalist-element)

**Comment by charlesakio** [comment]
We all agreed to use [DataList](https://developer.mozilla.org/en/docs/Web/HTML/Element/datalist)

**Comment by stemmlerjs** [comment]
well... We'll see about that^

**Comment by charlesjavelona** [comment]
Done
